seq2science/rules/configuration_generic.smk

```diff
@@ -107,27 +113,35 @@ for key, value in config.items():
 
 # read the samples.tsv file as all text, drop comment lines
 samples = pd.read_csv(config["samples"], sep='\t', dtype='str', comment='#')
-samples.columns = samples.columns.str.strip()
 
-# check that the columns are named
 assert all([col[0:7] not in ["Unnamed", ''] for col in samples]), \
     (f"\nEncountered unnamed column in {config['samples']}.\n" +
      f"Column names: {str(', '.join(samples.columns))}.\n")
 
-# check that the columns contains no irregular characters
-assert not any(samples.columns.str.contains('[^A-Za-z0-9_.\-%]+', regex=True)), \
-    (f"\n{config['samples']} may only contain letters, numbers and " +
-    "percentage signs (%), underscores (_), periods (.), or minuses (-).\n")
-
-# check that the file contains no irregular characters
-assert not any([any(samples[col].str.contains('[^A-Za-z0-9_.\-%]+', regex=True, na=False)) for col in samples if col != "control"]), \
-    (f"\n{config['samples']} may only contain letters, numbers and " +
-    "percentage signs (%), underscores (_), periods (.), or minuses (-).\n")
-
-# check that sample names are unique
-assert len(samples["sample"]) == len(set(samples["sample"])), \
-    (f"\nDuplicate samples found in {config['samples']}:\n" +
-     f"{samples[samples.duplicated(['sample'], keep=False)].to_string()}\n")
+column_schema = dict()
+allowed_pattern = r'[A-Za-z0-9_.\-%]+'
+column_schema["sample"] = [LeadingWhitespaceValidation(),
+                           TrailingWhitespaceValidation(),
+                           IsDistinctValidation(),
+                           MatchesPatternValidation(allowed_pattern)]
+
+column_schema["descriptive_name"] = [LeadingWhitespaceValidation(),
+                                     TrailingWhitespaceValidation(),
+                                     IsDistinctValidation(),
+                                     MatchesPatternValidation(allowed_pattern)]
```

Comment:
all columns are checked in the for loop after this, so maybe make a for loop for these two columns to check the `IsDistinceValidation()`?

Reply:
Yep makes sense!